Yosemite/YosemiteTests/Stores/ProductCategoryStoreTests.swift

```diff
@@ -194,6 +193,70 @@ final class ProductCategoryStoreTests: XCTestCase {
         XCTAssertNotNil(errorResponse)
     }
 
+    func testAddProductCategoryAddsStoredCategorySuccessfulResponse() {
```

Comment:
As I've shared in https://github.com/woocommerce/woocommerce-ios/pull/2326, we should really consider not using JSON files in `StoreTests`. In my experience, JSON files are just so difficult to debug and maintain:

- The data to test is in a separate file (the JSON) so we'll have to go back and forth to check the data that we're testing against. 
- The JSON file is already shared with the `RemoteTests`. If we change the JSON file, this `StoreTests` will most probably break too. I believe unit tests should be as encapsulated as possible. They should maintain their own "inputs" whenever possible.
- Using simple objects in `StoreTests` gives us more flexibility than JSON. It's easier to add different combinations of test data. 

So, my suggestion would be to do the same thing as https://github.com/woocommerce/woocommerce-ios/pull/2326.

### Add Protocol

```swift
public protocol ProductCategoriesRemoteProtocol {
    func loadAllProductCategories(for siteID: Int64,
                                         pageNumber: Int,
                                         pageSize: Int,
                                         completion: @escaping ([ProductCategory]?, Error?) -> Void)

    func createProductCategory(for siteID: Int64,
                                      name: String,
                                      parentID: Int64?,
                                      completion: @escaping (ProductCategory?, Error?) -> Void)
}

/// Product Categories: Remote Endpoints
///
public final class ProductCategoriesRemote: Remote, ProductCategoriesRemoteProtocol {
```

### Add Mock

```swift
final class MockProductCategoriesRemote: ProductCategoriesRemoteProtocol {
    private struct CreationResultKey: Hashable {
        let siteID: Int64
        let name: String
        let parentID: Int64?
    }

    private var creationResults = [CreationResultKey: Result<ProductCategory, Error>]()

    func loadAllProductCategories(for siteID: Int64, pageNumber: Int, pageSize: Int, completion: @escaping ([ProductCategory]?, Error?) -> Void) {
    }

    func createProductCategory(for siteID: Int64, name: String, parentID: Int64?, completion: @escaping (ProductCategory?, Error?) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                return
            }

            let key = CreationResultKey(siteID: siteID, name: name, parentID: parentID)
            if let result = self.creationResults[key] {
                switch result {
                case .success(let category):
                    completion(category, nil)
                case .failure(let error):
                    completion(nil, error)
                }
            } else {
                XCTFail("\(String(describing: self)) Could not find Result for \(key)")
            }
        }
    }

    func whenCreating(siteID: Int64, name: String, parentID: Int64?, thenReturn result: Result<ProductCategory, Error>) {
        let key = CreationResultKey(siteID: siteID, name: name, parentID: parentID)
        creationResults[key] = result
    }
}
```

We could probably make a generator for this in the future when it gets painful to create mocks. 

### Make Remote a Dependency in `ProductCategoryStore`

```swift

    private let remote: ProductCategoriesRemoteProtocol

    public override init(dispatcher: Dispatcher, storageManager: StorageManagerType, network: Network) {
        self.remote = ProductCategoriesRemote(network: network)
        super.init(dispatcher: dispatcher, storageManager: storageManager, network: network)
    }

    init(dispatcher: Dispatcher, storageManager: StorageManagerType, network: Network, remote: ProductCategoriesRemoteProtocol) {
        self.remote = remote
        super.init(dispatcher: dispatcher, storageManager: storageManager, network: network)
    }

    func addProductCategory(siteID: Int64, name: String, parentID: Int64?, onCompletion: @escaping (ProductCategory?, Error?) -> Void) {
        remote.createProductCategory(for: siteID, name: name, parentID: parentID) { [weak self] (productCategory, error) in

```

### Use the Mock in the Tests

```swift
    func testAddProductCategoryAddsStoredCategorySuccessfulResponse() throws {
        // Given a stubed product category network response
        let category = Yosemite.ProductCategory(categoryID: 104, siteID: sampleSiteID, parentID: 0, name: "Dress", slug: "Shirt")

        let remote = MockProductCategoriesRemote()
        remote.whenCreating(siteID: sampleSiteID, name: "Dress", parentID: 0, thenReturn: .success(category))

        let store = ProductCategoryStore(dispatcher: Dispatcher(), storageManager: storageManager, network: network, remote: remote)

        // When dispatching a `addProductCategory` action
        var errorResponse: Error?
        waitForExpectation { exp in
            let action = ProductCategoryAction.addProductCategory(siteID: sampleSiteID, name: "Dress", parentID: 0) { (productCategory, error) in
                errorResponse = error
                exp.fulfill()
            }

            store.onAction(action)
        }

        // Then the category should be added
        XCTAssertNil(errorResponse)

        let addedCategory = try XCTUnwrap(viewStorage.loadProductCategory(siteID: sampleSiteID, categoryID: 104))
        XCTAssertNotNil(addedCategory)
        XCTAssertEqual(addedCategory.categoryID, 104)
        XCTAssertEqual(addedCategory.parentID, 0)
        XCTAssertEqual(addedCategory.siteID, sampleSiteID)
        XCTAssertEqual(addedCategory.name, "Dress")
        XCTAssertEqual(addedCategory.slug, "Shirt")
    }
```

--- 

What do you think? 

Consider this just a discussion for now. It's fine if we don't have it in this PR. 🙂

Reply:
Thanks for your suggestions @shiki.

I'd say that ideally, I'd prefer a situation where we don't need to write a lot of code just to maintain our mock objects and a situation where a JSON file doesn't break everything because something changed. 😃 But sometimes there is no ideal and unique solution to things.

I think we could start using a mix of the two, maintaining JSON files for our Mapper class tests, and below I explain why I wouldn't completely abandon JSON files.
In our case, I found useful to have a JSON representation of the model, since they also include some fields which we don't parse at the moment, and the data follow the same type according to the backend rules.

Ofter happened to me to discovered that I was parsing a field in the wrong way, just because the JSON was returning an Int as String. In those specific cases, it will be useful to use a real JSON for parsing the data.

Another example is the use of real JSON files used for the `Product` model. There are some strange cases where some plugins alter a field value, for example, for the `price`, where the original value can be a String, an Int, or a Decimal. How would you handle this in a mock object which wasn't parsed from a real JSON?

Use simple objects can allow us to simplify how we use mocks in some cases, but testing every class Mapper can be a real pain, but I agree with you on using mock objects in `Store` tests. Another case is how we parse the fields which represent a date. Not going to use the correct date formatter and using objects that are not parsed, we risk getting false positives, but in reality, the date format has returned differently and we will notice it only when that field is used.

I know they are borderline cases, but having a real JSON copied and pasted from the documentation, or from a real network request, I think they make unit tests more useful and less free to interpretations. Ideally, a switch from a version X to a version Y of our API endpoints can be less painful, because copying the new JSON we could easily see which cases are failing, and what we need to update.

Do you think that make sense to use both approaches?

Reply:
> Do you think that make sense to use both approaches?

@pmusolino Yes and you're **absolutely right**. Thank you for sharing all that background. We'd definitely want the ability to test the cases that you mentioned. Using JSON files is probably the only way we can test that, aside from real API calls. 

I believe this is what we are both agreeing to: 

![image](https://user-images.githubusercontent.com/198826/83816936-c63afa00-a680-11ea-96e8-ee02258491d5.png)

- Remote Tests:
  - Test simulated API responses via JSON files
  - Test API errors
  - Endpoint and corresponding expected response testing
  - Some JSON parsing tests
- Mapper Tests:
  - Highly specialized JSON parsing testing
  - Test that we can _normalize_ the various and weird API responses 
- Store Tests:
  - Assumes that the Mapper and Remote Tests have fully covered the API response testing
  - Assumes that the responses from the (mocked) Remote are correctly parsed. There should be no "JSON parsing" testing here. 
  - Focused on business logic testing
  - Tests for database operations

How does this look? 

---

> for example, for the price, where the original value can be a String, an Int, or a Decimal. How would you handle this in a mock object which wasn't parsed from a real JSON?

If we follow the specs above, I believe this should be covered by the Mapper Tests. Or did I misunderstand this? 😅

Reply:
FYI @jaclync. In case you're interested in this discussion. 😄

Reply:
Great discussion! I like the plan above to test with the JSON files in Remote & Mapper tests, and a mock remote in Store tests 👍 

Testing with JSON simulates the API layer and allows us to test the handling of unexpected API response, so it makes the most sense to test with JSON in the Networking layer. I don't think we currently have many tests for unexpected JSON, but this JSON setup would be useful when we work on bug fixes. In the Yosemite layer, testing with a mock remote would be helpful to set the expected data more easily than with JSON (especially if we want to test different variations of the response).

Reply:
> How does this look?

Your proposal sounds good to me and makes totally sense.

> If we follow the specs above, I believe this should be covered by the Mapper Tests. Or did I misunderstand this?

This is correct 😃👍 

What do you think if we create a doc page with this information, just to not miss them anymore? What to do with all our existing test cases? Do we leave them as they are, and will we migrate them as soon as they need a makeover?

Reply:
> What do you think if we create a doc page with this information, just to not miss them anymore?

Yup. Thanks for the nudge. I created a task for it. https://github.com/woocommerce/woocommerce-ios/issues/2394

> What to do with all our existing test cases? Do we leave them as they are, and will we migrate them as soon as they need a makeover?

Yes, that's what I think. We can change them later if we really need to. For new test cases, we can start creating the mocks.

Reply:
A little late to this convo but I just saw it on the iOS-handbook.
I normally don't like json on unit tests because they are a pain to maintain, specially when we fail to update them and that gives us a sense of false safety.

In my ideal world we would be able to run our remote objects and mappers tests against some kind of local/test server that can serve JSONs from web/GH where we don't need to maintain them.

But since I've never accomplished that I'm happy to follow the current rules :-)

Reply:
> In my ideal world we would be able to run our remote objects and mappers tests against some kind of local/test server that can serve JSONs from web/GH where we don't need to maintain them.

I did this before. A downside to this is that the unit tests will no longer run without internet connection. Or if there are problems with the server, everyone gets blocked. 😬 

We do have some [Woo REST API unit tests](https://github.com/wordpress-mobile/WordPress-FluxC-Android/tree/develop/tests/api/src/test/java) as part of FluxC. That one runs on a real server I believe. We should get alerted if there's an API change that we did not expect. 

It's probably still useful to have a similar one fo the iOS side in the future. 

Or we could set up our mapper tests so that they can be tested against a local server, possibly using our variant of [VVV](https://github.com/Varying-Vagrant-Vagrants/VVV) that anyone can easily set up.

Reply:
Created https://github.com/woocommerce/woocommerce-ios/issues/2480 so the idea will not die.